### DISCUSSION TITLE
Node-redis v4 and node-redis v5 diffrences

### DISCUSSION THREAD

**Discussion by SeyyedKhandon** [NodeRedis/node_redis#2769] [Q&A] [answered]
Hey team,
I am going to use redis in my new project, but am not sure to go with `ioredis` for now or just using `node-redis v4` and then later when v5 got stable switch to it.
Is there any major differences on using the api's or breaking changes from version 4 to 5? where can I see these changes?

**Comment by leibale** [comment]
I suggest you start with v4 and upgrade to v5 whenever it's ready. See [this for breaking changes](https://github.com/redis/node-redis/blob/v5/docs/v4-to-v5.md) (we are still debating some of them), and [here this for "main" new features](https://github.com/redis/node-redis/blob/v5/docs/v5.md).

If you are using one of the commands that will break (for example, you are using EXISTS, which in v4 returns a boolean, but in v5 returns a number), you can either:
1. have a function that mimics the v5 reply (using `client.sendCommand` and use it instead for now).
2. use the old interface, and when you upgrade TypeScript should help you find all the places that broke..

**Comment by SeyyedKhandon** [comment, marked as answer]
Unfortunately for working with clusters, node-redis (`createCluster`) doesn’t provide the basic functionalities which it has on `createClient` such as status of the connection, flushall etc. which makes it very hard to work, in addition to that it comes with wiered errors that are no documented on the site or somewhere else, and the worst part is the type system which does not work properly for it.

Therefore I am going back to `ioredis`, I hope that `ClusterAllFailedError: Failed to refresh slots cache.` in ioredis gets fixed

**Reply by leibale** [reply]
@SeyyedKhandon I'm sorry you feel this way. Let me go over your points 1 by 1:
Regarding tracking the cluster connection status - this is indeed a limitation in v4 that is already "fixed" in v5.
Regarding `flushall` in a cluster -
In `ioredis`, calling `cluster.flushall` will execute the flushall command randomly in 1 of the cluster nodes (so it won't actually "flush all").
In node-redis we do not expose cluster.flushAll yet, but we plan to do it in v5.1 (but unlike ioredis, it'll follow redis command tips and execute it across the cluster). Until then you can use `cluster.masters` with `Promise.all` to execute `flushAll` across the cluster:
Regarding the weird undocumented errors - can you please share them so we can either document them or fix them?
"worst part is the type system which does not work properly for it" - can you please elaborate more on that?

**Reply by SeyyedKhandon** [reply]
Thank you very much for being very helpful, I greatly apperciate it.
Well for flushing, ioredis provides a better/simpler way tbh compared to node-redis, despite in both it requires to select masters and then send the command.
And for errors or other issues for example this  https://github.com/redis/node-redis/discussions/2774 is one of them. Basically v4 has lack some crucial functionalities(when I was working on the adaptor for it, I constantly was searching for each problem and found out various open issues which you mentioned inside of them that those issues or feature will be addressed in v5, and since  the candidate release date of v5 is not yet disclosed, so for us it makes sense to go for io-redis and then migrate from io-redis to node-redis v5.
And for type system io-redis has on outstanding type definition which is very helpful and the document also is again great, but for node-redis unfortunately some parts for example the cluster part of the document seems missing.
